### PR TITLE
New lint [`bare_dos_device_names`]

### LINKED ISSUE #8511: Warn on use of Windows DOS devices in paths
### What it does

If the user uses a bare dos device name such as `aux`, windows will convert it to `\\.\aux` (the aux device). This may be surprising. Clippy could warn the user and suggest they explicitly use `\\.\aux` if that was intended. Using `.\aux` could also suppress this lint.

The DOS device names are (case insensitive):

* AUX
* CON
* CONIN$
* CONOUT$
* COM1, COM2, COM3, COM4, COM5, COM6, COM7, COM8, COM9, COM², COM³, COM¹
* LPT1, LPT2, LPT3, LPT4, LPT5, LPT6, LPT7, LPT8, LPT9, LPT², LPT³, LPT¹
* NUL
* PRN

### Lint Name

bare_dos_device_name

### Category

suspicious

### Advantage

Avoids unexpected and surprising path transformations.

### Drawbacks

This is platform specific behaviour.

### Example

```rust
File::open("aux");
```

Could be written as:

```rust
File::open("\\.\aux");
```

Or 

```rust
File::open(".\aux");
```

### PR DESCRIPTION
**Pull request by Centri3**
Closes #8511

I've made this a draft as it's not quite ready yet. But whoever's randomly picked, feel free to review ^^ Everything I've left out so far and/or haven't finished is really because I didn't get enough sleep :D I've left some TODOs about the place and will likely make it a tad more general soon enough.

changelog: New lint [`bare_dos_device_names`]

### CODE REVIEW

File: tests/ui/bare_dos_device_names.stderr [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+error: this path refers to a DOS device
+  --> $DIR/bare_dos_device_names.rs:19:16
+   |
+LL |     File::open("conin$");
+   |                ^^^^^^^^
+   |
+help: if this is intended, try
+   |
+LL |     File::open("//./conin$");
```
Comment (Centri3) [inline tests/ui/bare_dos_device_names.stderr:40]: Those are supposed to be backslashes :sweat_smile:

Comment (Centri3) [reply tests/ui/bare_dos_device_names.stderr:40]: It appears `span_suggestion_verbose` (and the others) all normalize strings... I'm not sure how to fix this :confused: I saw other cases where lints use `\` but I can't figure out how to make it not convert it to `/` :/

**Comment by xFrednet** [comment]
Hey this is triage, I'm closing this due to inactivity. Currently, @Centri3 sadly doesn't have the time to continue this implementation. If anyone is interested in continuing this PR, you're more than welcome to create a new PR and push it over the finish line. :D

Thank you to @Centri3 and the reviewers for the time, that you already put into this!

@rustbot label +S-inactive-closed -S-waiting-on-author -S-waiting-on-review

**Comment by Centri3** [comment]
Triaging my old PRs in case someone wants to pick them up, I don't really think the implementation needs any more work, but there is a TODO left as a question - Should we lint these specific device names?

Regardless I think I'll get around to this too at some point if nobody else does

**Pull request closed without merging**